Update contributing guide (#1827)
vdaas/vald: CONTRIBUTING.md

# Contributing to Vald

Thank you for your interest in Vald, and thank you for investing your time in contributing to Vald!

Please read our [Code Of Conduct](https://github.com/vdaas/vald/blob/main/CODE_OF_CONDUCT.md) to keep our community approachable and respectable.

In this guide, you will get an idea of how to contribute to Vald.

If you are not a developer, don't worry, some contributions don't require writing a single line of code.

## Contributions

> Please read the [README](https://github.com/vdaas/vald/blob/main/README.md) to get an overview of Vald.

Reviewer: ```suggestion
Please read the [About Vald](https://vald.vdaas.org/docs/overview/about-vald) to get an overview of Vald.
```
:thinking:

We welcome you to contribute to Vald to make Vald better.
We accept the following types of contributions:

- Issue
  - Bug report
  - Feature request

>   - Proposal

Reviewer: what is the difference between a feature request and a proposal?
Sometimes, it has the same meaning.

Author: ```suggestion
  - Feature request / Proposal
```

  - Security issue report

> - Pull request

Reviewer: You should break down the type of `Pull Request`.
It has many types, which have each original step for contributing.

Author: ```suggestion
- Pull request
  - Source code implementation
  - Test implementation
  - Documentation
```

Please also feel free to ask anything on [Vald Slack channel](https://join.slack.com/t/vald-community/shared_invite/zt-db2ky9o4-R_9p2sVp8xRwztVa8gfnPA) :)

## Issue contribution

> Please find [here](./docs/contributing/issue.md) for more details about how to contribute to the issue.

Reviewer: Too short.
At least, we have to introduce the `issue contribution` for the Vald project.

> ## Code contribution

Reviewer: You have to use the same word. Line 23.

Author: ```suggestion
## Pull request contribution
```

> ### Development environment installation guide

Author: ```suggestion
### Installation guide
```

#### Mac/Windows users

For Mac / Windows users, please install and use [docker](https://www.docker.com/) to create a Linux container to build and test Vald.

You need to install the packages and execute the commands listed in the [Linux section](#linux-users) inside the docker container rather than in your local environment.

If you want to start development on your host environment rather than in the docker container, please consider mounting the Vald repository path from the host environment to the container environment after cloning the Vald repository described below.

```bash
docker run -v '{vald repo}':'{folder mount}' {container name}
```

For more details about docker, please refer to the [docker documentation](https://docs.docker.com/get-started/overview/).

#### Linux users

For Linux users, please install the following tools on your environment.

- [curl](https://curl.se/)
- [make](https://www.gnu.org/software/make/)
- [cmake](https://cmake.org/)
- [Protobuf](https://grpc.io/docs/protoc-installation/)
- [npm](https://www.npmjs.com/)
- [unzip](https://linux.die.net/man/1/unzip)
- [Git](https://git-scm.com/)

> - [Go](https://go.dev/) (v1.19.2 is recommended)

Reviewer: ```suggestion
- [Go](https://go.dev/) (v1.19.2 or later is recommended)
```

make tools/install # install development tools like helm, kind, etc.
make gotests/install # install gotests tools to generate test stubs.
```

#### Before making changes

Before making changes on Vald, please follow these steps to contribute to any of our open-source repositories:

1. Ensure that you have completed our [CLA Agreement](https://cla-assistant.io/vdaas/vald)
2. Set your name and email (these should match the information on your submitted CLA)

    ```bash
    git config --global user.name "Firstname Lastname"
    git config --global user.email "your_email@example.com"
    ```

    Please also refer [here](https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup) for more details on setting up Git.
3. Setup signing key on your development environment
    Please refer [here](https://docs.github.com/authentication/managing-commit-signature-verification/telling-git-about-your-signing-key) to configure the signing key.
    Vald recommends signing the commit to prove that the commit actually came from you, as it is easy to add anyone as an author of the commit, which can be used in hiding the author of malicious code.

#### Make changes

> ##### 1. Make sure no one is working on the same issue/feature

Reviewer: ```suggestion
1. Make sure no one is working on the same issue/feature
```

Before making any changes, you need to check if anyone is working on the same feature in the pull request list.

If you are solving an issue, check if anyone is working on the issue and comment on the issue and say you are working on it to avoid conflict with others also working on the same issue.

> ##### 2. Create your feature branch on your forked repository

Reviewer: ```suggestion
2. Create your feature branch on your forked repository
```

| Field       | Explanation                           | Naming Rule                                                                                                                 |
| :---------- | :------------------------------------ | :-------------------------------------------------------------------------------------------------------------------------- |
| type        | The PR type                           | The type of PR can be a feature, bug, refactoring, benchmark, security, documentation, dependencies, ci, test, etc...       |
| area        | Area of context                       | The area of PR can be gateway, agent, agent-sidecar, lb-gateway, etc...                                                     |
| description | Summarized description of your branch | The description must be hyphenated. Please use [a-zA-Z0-9] and a hyphen as characters, and do not use any other characters. |

(\*) If you changed multiple areas, please list each area with "-".

For example, when you add a new feature for internal/servers, the name of the branch will be `feature/internal/add-newfeature-for-servers`.

```bash
git checkout -b [type]/[area]/[description]
```

> ##### 3. Make code changes

Reviewer: ```suggestion
3. Make code changes
```

If you have discussed the design and the requirement of the changes with Vald members, please follow the design and requirement discussed.

To make changes on Vald, you need to understand the package structure of Vald and know which part you need to make changes.

| Package name | Description                                                      |
| :----------- | :--------------------------------------------------------------- |
| apis         | Vald API definitation                                            |
| cmd          | Entry point of the Vald components                               |
| example      | Example code of Vald                                             |
| internal     | Internal package to extend and customize libraries functionality |
| pkg          | Contains business logic implementation                           |

Please make sure you understand our [coding guideline](./docs/contributing/coding-style.md) to follow our coding style to keep the coding style consistent.

> ##### 4. After making code changes

Reviewer: ```suggestion
5. After making code changes
```

Author: ```suggestion
4. After making code changes
```

After making code changes, we suggest you execute the following command to generate the necessary test stubs and format code.

```bash
make gotests/gen # execute gotests tools to generate unit test code stubs
make format # format go and yaml files
```

The command `make gotests/gen` generate unit test code stubs to easier to implement unit test code.

The command `make format` is used to generate the license header on the source code file, and execute the code formatter to format Go and YAML files.

It will also install the following tools to format the source code.

- [golines](https://github.com/segmentio/golines)
- [gofumpt](https://github.com/mvdan/gofumpt)
- [strictgoimports](https://github.com/momotaro98/strictgoimports)
- [goimports](https://pkg.go.dev/golang.org/x/tools/cmd/goimports)
- [prettier](https://prettier.io/)

These tools are required to format your source code to keep the coding style consistent.

> ##### 5. Verify the changes

Reviewer: ```suggestion
5. Verify the changes
```

If you are making logical changes on Vald, please refer to [this document](./docs/contributing/testing-guideline.md) for more detail about how to test your changes.

> ##### 6. Commit and push your changes to the branch

Reviewer: ```suggestion
6. Commit and push your changes to the branch
```

After verifing the changes, you may want to push the changes to your development branch.

Please add the files that related to your changes only.

```bash
git add [files]
```

Please write a brief description of the changes to the commit, and push it to your forked repository.

```bash
git commit --signoff -m '[commit message]'
git push origin [type]/[area]/[description]
```

> ##### 7. Create a new pull request against the Vald repository

Reviewer: ```suggestion
7, Create a new pull request against the Vald repository
```

Author: ```suggestion
7. Create a new pull request against the Vald repository
```

| type/dependency  | For dependency update pull request   |
| type/feature     | For new feature pull request         |
| type/refactoring | For code refactoring pull request    |
| type/security    | For security fix pull request        |
| type/test        | For test implementation pull request |

We also provide the following label to execute specific actions on the [GitHub Actions](https://github.co.jp/features/actions).

| Label                | Description                                           |
| :------------------- | :---------------------------------------------------- |
| action/e2e-chaos     | Execute E2E chaos test                                |
| action/e2e-deploy    | Execute E2E deployment test                           |
| action/e2e-max-dim   | Execute maximum dimension E2E test                    |
| action/e2e-profiling | Execute E2E test with profiling                       |
| action/fossa         | Execute [fossa](https://fossa.com/) security checking |

Please add the action label to the pull request to execute specific action if needed.

If you are solving an issue, please also link the pull request to the issue.

> ##### 8. Review and merge pull request

Reviewer: ```suggestion
8. Review and merge pull request
```

# Contributing to Vald

Thank you for your interest in Vald, and thank you for investing your time in contributing to Vald!

Please read our [Code Of Conduct](https://github.com/vdaas/vald/blob/main/CODE_OF_CONDUCT.md) to keep our community approachable and respectable.

In this guide, you will get an idea of how to contribute to Vald.

If you are not a developer, don't worry, some contributions don't require writing a single line of code.

## Contributions

Please read the [About Vald](https://vald.vdaas.org/docs/overview/about-vald) to get an overview of Vald.

We welcome you to contribute to Vald to make Vald better.
We accept the following types of contributions:

- Issue
  - Bug report
  - Feature request / Proposal
  - Security issue report
- Pull request

>   - Source code implementation
>     - Business logic implementation
>   	- Test implementation
> 	- Documentation

Author: ```suggestion
  - Source code implementation
    - Business logic implementation
    - Test implementation
  - Documentation
```

Please also feel free to ask anything on [Vald Slack channel](https://join.slack.com/t/vald-community/shared_invite/zt-db2ky9o4-R_9p2sVp8xRwztVa8gfnPA) :)

## Issue contribution

[Issues](https://github.com/vdaas/vald/issues) are used to track tasks that contributors can help with.

Please find [here](./docs/contributing/issue.md) for more details about how to contribute to the issue.

## Pull request contribution

Pull request is also called merge request to let other knows about the changes you have made, to review and discuss the changes and finally merge it to the main branch.

> In Vald, you need to create an pull request to ask for the review and actually make changes to Vald.

Author: ```suggestion
In Vald, you need to create a pull request to ask for the review and actually make changes to Vald.
```

You need to create pull request to make changes on:
1. Source code changes
   1. Business logic implementation
   2. Test implementation
2. Documentation creation / update

> ### Install and configure Github

Author: ```suggestion
### Install and configure GitHub
```

> Before createing an pull request, you need to install and configure Github to create development branch, make changes and finally create a pull request.

Author: ```suggestion
Before creating a pull request, you need to install and configure Github to create development branch, make changes and finally create a pull request.
```

Please install [Git](https://git-scm.com/) and configure it first.

1. Ensure that you have completed our [CLA Agreement](https://cla-assistant.io/vdaas/vald).
1. Set your name and email (these should match the information on your submitted CLA).

    ```bash

>     git config --global user.name "Github user name"

Author: ```suggestion
    git config --global user.name "GitHub user name"
```

#### Windows / Mac

For Mac / Windows users, please download and install [docker](https://www.docker.com/) to create a Linux container to build and test Vald.

You need to install the packages and execute the commands listed in the [Linux section](#linux-users) inside the docker container rather than in your local environment.

If you want to start development on your host environment rather than in the docker container, please consider mounting the Vald repository path from the host environment to the container environment after cloning the Vald repository described below.

```bash
docker run -v '{vald repo}':'{folder mount}' {container name}
```

For more details about docker, please refer to the [docker documentation](https://docs.docker.com/get-started/overview/).

#### Linux

For Linux users, please install the following tools on your environment.

- [curl](https://curl.se/)
- [make](https://www.gnu.org/software/make/)
- [cmake](https://cmake.org/)
- [Protobuf](https://grpc.io/docs/protoc-installation/)
- [npm](https://www.npmjs.com/)
- [unzip](https://linux.die.net/man/1/unzip)

> - [Go](https://go.dev/) (v1.19.2 or later is recommended)

Author: ```suggestion
- [Go](https://go.dev/) (v1.19.5 or later is recommended)
```

| Run e2e {target}                         | Execute End-to-end testing                            |
| Run Helm lint / {chart name}             | Execute linter for Helm chart files                   |
| reviewdog - {target}                     | Execute reviewdog to review the changes on the target |
| DeepSource - {target}                    | Execute DeepSource to analysis target files           |

About the `Build docker image` pipeline, it will build the docker image for the specific tag and upload it to the [DockerHub](https://hub.docker.com/u/vdaas).

Specifically, it builds the docker image on every pull request and tags the image as `pr-{pull request number}`.

Whenever a pull request is merged into the main branch, a nightly build will be built and uploaded to the DockerHub.

When Vald release, this pipeline will build the docker image with a specific version (e.g. `v1.6.1`) and also update the `latest` tag image.

About the `Run e2e` pipeline, it will execute the end-to-end testing and fails if any error occurred.

This pipeline will be executed only if the label `e2e/{target}` is added to the pull request.

> Please note that if you are working in a forked repository, all of the CI/CD pipelines will not be executed automatically. You may need to ask the Vald owner to grant access by adding the `ci/approved` label to the pull request to execute the pipelines.

Author: ```suggestion
Please note that if you are working in a forked repository, all the CI/CD pipelines will not be executed automatically. You may need to ask the Vald owner to grant access by adding the `ci/approved` label to the pull request to execute the pipelines.
```

Read our [Code Of Conduct](https://github.com/vdaas/vald/blob/main/CODE_OF_CONDUCT.md) to keep our community approachable and respectable.

In this guide, you will get an idea of how to contribute to Vald.

If you are not a developer, don't worry, some contributions don't require writing a single line of code.

## New contributor guide

Please read the [README](https://github.com/vdaas/vald/blob/main/README.md) to get an overview of Vald.

We welcome you to contribute to Vald to make Vald better. We accept the following types of contributions:

- Issue
  - Bug report
  - Feature request
  - Proposal
  - Security issue report
- Pull request

Please also feel free to ask anything on [Vald Slack channel](https://join.slack.com/t/vald-community/shared_invite/zt-db2ky9o4-R_9p2sVp8xRwztVa8gfnPA) :)

If you decided to contribute source code changes, you may need to install the following tools manually before making changes.

- [git](https://git-scm.com/)
- [go](https://go.dev/)

Also, you may need to run the following command under [Vald repository](https://github.com/vdaas/vald) to install the necessary packages.

```bash
make init # initialize development environment

> make tools/install # install development tools like helm, kind, etc

Author: ```suggestion
make tools/install # install development tools like helm, kind, etc.
```
